src/OracleService/OracleService.cs

```diff
@@ -0,0 +1,481 @@
+        private Timer timer;
+        private readonly CancellationTokenSource cancelSource = new CancellationTokenSource();
+        private bool started = false;
+        private bool stopped = false;
+        private int counter;
+
+        private static readonly IReadOnlyDictionary<string, IOracleProtocol> protocols = new Dictionary<string, IOracleProtocol>
+        {
+            ["https"] = new OracleHttpsProtocol()
+        };
+
+        public override string Description => "Built-in oracle plugin";
+
+        public OracleService()
+        {
+            RpcServerPlugin.RegisterMethods(this);
+        }
+
+        protected override void Configure()
+        {
+            Settings.Load(GetConfiguration());
+            foreach (var p in protocols)
+                p.Value.Configure();
+        }
+
+        public override void Dispose()
+        {
+            OnStop();
+            while (!stopped)
+                Thread.Sleep(100);
+            foreach (var p in protocols)
+                p.Value.Dispose();
+        }
+
+        [ConsoleCommand("start oracle", Category = "Oracle", Description = "Start oracle service")]
+        private void OnStart()
+        {
+            if (started) return;
+            string password = GetService<ConsoleServiceBase>().ReadUserInput("password", true);
+            if (password.Length == 0)
+            {
+                Console.WriteLine("Cancelled");
+                return;
+            }
+
+            wallet = new NEP6Wallet(Settings.Default.Wallet);
+            try
+            {
+                wallet.Unlock(password);
+            }
+            catch (System.Security.Cryptography.CryptographicException)
+            {
+                Console.WriteLine($"Failed to open wallet");
+                return;
+            }
+
+            using (var snapshot = Blockchain.Singleton.GetSnapshot())
+            {
+                if (!CheckOracleAvaiblable(snapshot, out ECPoint[] oracles)) throw new ArgumentException("The oracle service is unavailable");
+                if (!CheckOracleAccount(wallet, oracles)) throw new ArgumentException("There is no oracle account in wallet");
+            }
+
+            started = true;
+
+            timer = new Timer(OnTimer, null, RefreshInterval, Timeout.Infinite);
+
+            ProcessRequestsAsync();
+        }
+
+        [ConsoleCommand("stop oracle", Category = "Oracle", Description = "Stop oracle service")]
+        private void OnStop()
+        {
+            cancelSource.Cancel();
+            if (timer != null)
+            {
+                timer.Dispose();
+                timer = null;
+            }
+        }
+
+        void IPersistencePlugin.OnPersist(StoreView snapshot, IReadOnlyList<Blockchain.ApplicationExecuted> applicationExecutedList)
+        {
+            if (!CheckOracleAvaiblable(snapshot, out ECPoint[] oracles) || !CheckOracleAccount(wallet, oracles))
+                OnStop();
```

Comment:
this will set cancelSource.IsCancellationRequested as true, when next time start oracle, it won't be able to create oracle response.

Reply:
If the service is stopped, we should not reuse the service for the next start.

Reply:
Since before starting oracle, OnPersist will be executed, and wallet is null which will lead to CheckOracleAccount return false and then be forced to stop. 

```
        private static bool CheckOracleAccount(Wallet wallet, ECPoint[] oracles)
        {
            if (wallet is null) return false;
            return oracles
                .Select(p => wallet.GetAccount(p))
                .Any(p => p is not null && p.HasKey && !p.Lock);
        }
```

Reply:
Add `if (!started) return;`?